File: utils/logger.py
import logging
from abc import ABC, abstractmethod
from typing import Any, Dict, Optional

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)


class Logger(ABC):
    """Abstract base class for a logger."""

    @abstractmethod
    def error(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log an error message."""
        pass

    @abstractmethod
    def info(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log an info message."""
        pass

    @abstractmethod
    def warning(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log a warning message."""
        pass

    @abstractmethod
    def debug(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log a debug message."""
        pass


class BasicLogger(Logger):
    """A basic implementation of the Logger interface using Python's built-in logging module."""

    def error(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log an error message with optional context."""
        if context:
            logging.error(f"{message} | Context: {context}")
        else:
            logging.error(message)

    def info(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log an info message with optional context."""
        if context:
            logging.info(f"{message} | Context: {context}")
        else:
            logging.info(message)

    def warning(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log a warning message with optional context."""
        if context:
            logging.warning(f"{message} | Context: {context}")
        else:
            logging.warning(message)

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log a debug message with optional context."""
        if context:
            logging.debug(f"{message} | Context: {context}")
        else:
            logging.debug(message)

File: src/utils/logger.ts
/**
 * Logger interface and implementation.
 */

export interface Logger {
  error(message: string, context?: Record<string, unknown>): void;
  info(message: string, context?: Record<string, unknown>): void;
  warning(message: string, context?: Record<string, unknown>): void;
  debug(message: string, context?: Record<string, unknown>): void;
  enabled: boolean;
}

export class BasicLogger implements Logger {
  private _enabled: boolean = true;

  get enabled(): boolean {
    return this._enabled;
  }

  set enabled(value: boolean) {
    this._enabled = value;
  }

  error(message: string, context?: Record<string, unknown>): void {
    if (!this._enabled) return;
    const timestamp = new Date().toISOString();
    const contextStr = context ? ` | Context: ${JSON.stringify(context)}` : '';
    console.error(`${timestamp} - ERROR - ${message}${contextStr}`);
  }

  info(message: string, context?: Record<string, unknown>): void {
    if (!this._enabled) return;
    const timestamp = new Date().toISOString();
    const contextStr = context ? ` | Context: ${JSON.stringify(context)}` : '';
    console.log(`${timestamp} - INFO - ${message}${contextStr}`);
  }

  warning(message: string, context?: Record<string, unknown>): void {
    if (!this._enabled) return;
    const timestamp = new Date().toISOString();
    const contextStr = context ? ` | Context: ${JSON.stringify(context)}` : '';
    console.warn(`${timestamp} - WARNING - ${message}${contextStr}`);
  }

  debug(message: string, context?: Record<string, unknown>): void {
    if (!this._enabled) return;
    const timestamp = new Date().toISOString();
    const contextStr = context ? ` | Context: ${JSON.stringify(context)}` : '';
    console.debug(`${timestamp} - DEBUG - ${message}${contextStr}`);
  }
}
